homeassistant/components/eq3btsmart/climate.py

```diff
@@ -186,14 +197,26 @@ def set_preset_mode(self, preset_mode):
         """Set new preset mode."""
         if preset_mode == PRESET_NONE:
             self.set_hvac_mode(HVAC_MODE_HEAT)
-        self._thermostat.mode = HA_TO_EQ_PRESET[preset_mode]
+
+        self._BTUpdateLock.acquire()
+
+        try:
+            self._thermostat.mode = HA_TO_EQ_PRESET[preset_mode]
+        finally:
+            self._BTUpdateLock.release()
 
     def update(self):
         """Update the data from the thermostat."""
         # pylint: disable=import-error,no-name-in-module
         from bluepy.btle import BTLEException
 
+
+        self._BTUpdateLock.acquire()
```

Comment:
We normally don't allow sync locks since they risk exhausting the thread pool. An async lock would be ok, but that requires switching to use the async home assistant api for all methods that need to access the lock. Regular I/O will then need to be explicitly scheduled on the executor thread pool.

Reply:
Thanks for your comment. You mean asyncio?

Reply:
Yes. See eg the fibaro light platform:
https://github.com/home-assistant/home-assistant/blob/15bb12f48eddf2416519f6568da1f1b118aeb533/homeassistant/components/fibaro/light.py#L181-L187

Reply:
@MartinHjelmare the core don't call all function at once like on first async implementation. So it would be save to use lock also inside executor because they are not querying multiple times. @balloob ?

Reply:
If an update is in progress, and an automation with retry calls an actuator service we could consume multiple threads, I'm thinking.

Reply:
So Martin is right, if I execute a 100 scripts, they will all in parallel hit this lock and we can starve the threadpool.

The actual update of the entity is guarded with a semaphore, but that is not impacting services that are called.

Reply:
Here is an async-safe version of the set temperature method.

```python
    async def async set_temperature(self, **kwargs):
        """Set new target temperature."""
        temperature = kwargs.get(ATTR_TEMPERATURE)
        if temperature is None:
            return

        async with self._BTUpdateLock:
            await self.hass.async_add_executor_job(setattr, 'target_temperature', temperature)
```

Reply:
Thx for hints I will get back to this pull request soon. One more concern about lock in python. Let’s say we have a case of two automation and background update in progress.
Update thread get a lock and other threads need to wait. I believe we should assure that we will apply setting in the same order as it was called by automation or user. 
Let’s say one automation is setting temperature to 21 degrees and 10 seconds after another sets it to 15 degrees. Both automations threads are waiting on lock it is important which will be allowed first. 
Is lock or async mechanism in Python can assure chronological order or I should implement queue for this?

Reply:
Chronological order is ensured.